File: workshop/cpp/events/template.cpp
// https://stackoverflow.com/a/66946587

#include <cassert>
#include <string_view>
#include <utility>
#include <vector>
#include <fmt/core.h>

#include "elf/elf.hpp"

auto solve_first_part(std::string_view filepath) -> void {
  std::string solution =
      fmt::format("Unimplemented. No solution for {}", filepath);
  elf::print_solution(1, solution);
}

auto solve_second_part(std::string_view filepath) -> void {
  std::string solution =
      fmt::format("Unimplemented. No solution for {}", filepath);
  elf::print_solution(2, solution);
}

auto main(const int argc, const char *argv[]) -> int {
  const std::vector<std::string_view> args(
      argv, std::next(argv, static_cast<ptrdiff_t>(argc)));

  const auto config = elf::parse_command_line(args);

  for (const auto part : config.parts) {
    if (part == 1) {
      solve_first_part(config.input);
    } else if (part == 2) {
      solve_second_part(config.input);
    } else {
      std::unreachable();
    };
  }
}

File: solutions/python/events/template.py
from pathlib import Path
from typing import assert_never

import elf


def solve_first_part(__p: Path, /) -> None:
    solution = f"Unimplemented. No solution for {__p!s}"
    elf.print_solution(1, solution)


def solve_second_part(__p: Path, /) -> None:
    solution = f"Unimplemented. No solution for {__p!s}"
    elf.print_solution(2, solution)


def main() -> None:
    parts, input = elf.parse_command_line()
    for part in parts:
        match part:
            case 1:
                solve_first_part(input)
            case 2:
                solve_second_part(input)
            case _:
                assert_never(part)


if __name__ == "__main__":
    main()
